src/Consumer.cc

```diff
@@ -79,144 +81,131 @@ void MessageListener(pulsar_consumer_t *cConsumer, pulsar_message_t *cMessage, v
   listenerCallback->callback.Release();
 }
 
-void Consumer::SetCConsumer(std::shared_ptr<CConsumerWrapper> cConsumer) { this->wrapper = cConsumer; }
-void Consumer::SetListenerCallback(ListenerCallback *listener) {
+void Consumer::SetCConsumer(std::shared_ptr<pulsar_consumer_t> cConsumer) { this->cConsumer = cConsumer; }
+void Consumer::SetListenerCallback(MessageListenerCallback *listener) {
   if (listener) {
+    // Pass consumer as argument
+    listener->consumer = this;
+  }
+
+  if (this->listener == nullptr) {
     // Maintain reference to consumer, so it won't get garbage collected
     // since, when we have a listener, we don't have to maintain reference to consumer (in js code)
     this->Ref();
```

Comment:
Why increase the reference count when `listener` is `nullptr`?

Reply:
FYI: https://github.com/apache/pulsar-client-node/pull/83

Reply:
I think this was purely to make sure that the ref count is balanced. So if we set the listener 10 times, we don't Ref the Consumer more than once, as there is only one corresponding Unref in `Consumer::Cleanup()`.

Reply:
Is it correct that the reference counter is incremented even if the user does not set the message listener?

Reply:
The reference counter should not be incremented if the message listener is not set. This only gets called in the constructor of the Consumer if there is a listener. All this change does is ensure that if someone were to call this in future from somewhere else the references would remain balanced. I will add a check to make sure that the listener being set isn't itself nullptr. The change just makes the code more resilient rather than changing any actual behaviour.